torchtitan/components/checkpoint.py

```diff
@@ -134,15 +151,78 @@ def checkpoint_mp(recv, send):
             state, checkpoint_id = obj
             save_with_gc(state, checkpoint_id=checkpoint_id)
             logger.info(
-                "Finish saving the checkpoint in the background process in "
-                f"{time.monotonic() - begin:.2f} seconds."
+                "Finish saving the checkpoint in the background process in %.2f seconds.",
+                time.monotonic() - begin,
             )
     finally:
         logger.info("Destroying the process group.")
         dist.destroy_process_group()
 
 
+def purge_thread(purge_queue: queue.Queue):
+    """Thread to purge the old checkpoints.
+
+    This is only used when keep_latest_k > 0.
+
+    Args:
+        purge_queue (queue.Queue): The queue to receive the path to purge and Terminate signal.
+    """
+    try:
+        while True:
+            path = purge_queue.get()
+            if isinstance(path, Terminate):
+                return
+            assert isinstance(path, str)
+            logger.info("Checkpointer is deleting %s.", path)
+            begin = time.monotonic()
+            shutil.rmtree(path, ignore_errors=True)
+            logger.info(
+                "Checkpointer deleted %s in %.2f seconds.",
+                path,
+                time.monotonic() - begin,
+            )
+    finally:
+        logger.info("Destroying the purge thread.")
+
+
 class CheckpointManager:
+    """This class manages the checkpointing logic for the TorchTitan trainer.
+
+
+    Note: Pipeline Parallelism and Virtual Stages
+
+    1. even for simple PP schedules, there is a separate optimizer each PP rank.
+    rank0's optimizer would have a param_group[0] which refers to layers.0 in the original
+    model.  rank1's would _also_ have a param_group[0], since it's index based, but
+    referring to layers.1.  When saving, these collide and one of them is lost.  Then when
+    reloading, only one stage can restore its optimizer states, others will error.
+
+        The solution to this problem is optimizer flattening: it landed in #127071 and is
+        enabled in TorchTitan by passing the 'flatten_optimizer_state_dict' kwarg to DCP
```

Comment:
Nit: not needed in this PR. It would be nice, e.g., if we can give one example, like how flatten works in this case, that would be great.